src/preferences/common/preference-service.ts

```diff
@@ -0,0 +1,133 @@
+// import { inject, injectable } from "inversify";
+import { FileSystem, FileChangesEvent } from "../../filesystem/common/filesystem";
+// import * as path from "path";
+// import { FileSystemNode } from "../../filesystem/node/node-filesystem";
+
+
+
+import { FileSystemWatcher } from "../../filesystem/common/filesystem-watcher"
+
+export class Preference {
+    constructor(public key: string, public value: any) { }
+}
+export interface PreferenceChangedEvent {
+    readonly preferenceName: string;
+    readonly newValue?: any;
+    readonly oldValue?: any;
+}
+
+/**
+ * Minimal preference API with basic functionalities
+ */
+export interface PreferenceService {
+    // A minimal part of the interface.
+    // readonly preferenceNames: Promise<string[]>;
+
+    has(preferenceName: string): Promise<boolean>;
+    get<T>(preferenceName: string): Promise<T | undefined>;
+    // get<T>(preferenceName: string, defaultValue?: T): Promise<T>;
+    // set<T>(preferenceName: string, value: T): void;
+    // remove<T>(preferenceName: string): T | undefined;
+
+}
+
+export interface PreferenceServiceClient {
+    onDidChangePreference(event: PreferenceChangedEvent): void
+}
+
+export class JsonPreferenceService implements PreferenceService {
+
+    // A minimal part of the interface.
+    protected prefs: { [key: string]: any } | undefined; // Preferences cache
+    protected readonly prefsChangedListeners: Emitter<PreferenceChangedEvent> = new Emitter();
+    protected readonly resolveUri: Promise<URI>;
+
+    constructor(
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
+        readonly preferencePath: string) {
+
+        this.resolveUri = fileSystem.getWorkspaceRoot().then(root =>
+            new URI(root.uri).appendPath(preferencePath)
+        );
+
+
+
+        watcher.onFileChanges(event => {
+            this.arePreferencesAffected(event).then((arePrefsAffected) => {
+                if (arePrefsAffected) {
+                    this.reconcilePreferences();
+                }
+            })
+        });
+
+        this.reconcilePreferences();
+    }
+
+    /**
+     * Checks to see if the preference file was modified
+     */
+    arePreferencesAffected(event: FileChangesEvent): Promise<boolean> {
+        return this.resolveUri.then(uri =>
+            event.changes.some(c =>
+                c.uri === uri.toString()
+            )
+        );
+    }
+
+    /**
+     * Read preferences
+     */
+    reconcilePreferences(): void {
+
+        this.resolveUri.then((uri) => {
+            // const uri = path.resolve(root, "/", this.preferencePath);
+            this.fileSystem.resolveContent(uri.toString()).then(({ stat, content }) => {
+                let newPrefs = JSON.parse(content) // Might need a custom parser because comments and whatnot?
+                const newKeys: { [key: string]: any } = Object.keys(newPrefs).sort();
```

Comment:
I had to write this type as I couldn't find a way to get around the following error:


```
'Element implicitly has an 'any' type because index expression is not of type 'number'.'
at: '95,41'
```

And yes I think the sorts are useless.